File: frontend/src/lib/api/endpoints/ai.ts
import { ApiClient } from '../client';
import type {
	AiStatus,
	AiAgentListResponse,
	AiRunListResponse,
	AiAnswerRequest,
	AiAnswerResponse,
	AiModelListResponse,
	AiModelRoute,
	AiModelRouteUpdateRequest,
	AiMeetingPrepRequest,
	AiMeetingPrepResponse,
	AiPromptActivateRequest,
	AiPromptCreateRequest,
	AiPromptEvalRun,
	AiPromptListResponse,
	AiPromptTemplate,
	AiPromptTestRequest,
	AiPromptVersion,
	AiPromptVersionCreateRequest,
	AiProviderAccount,
	AiProviderCommandResponse,
	AiProviderConsentRequest,
	AiProviderCreateRequest,
	AiProviderListResponse,
	AiProviderPatchRequest,
	AiSettingsOverviewResponse
} from '../types';

export async function fetchAiStatus(): Promise<AiStatus> {
	return ApiClient.instance.get<AiStatus>('/api/v1/ai/status', 'AI status request failed');
}

export async function fetchAiAgents(): Promise<AiAgentListResponse> {
	return ApiClient.instance.get<AiAgentListResponse>('/api/v1/ai/agents', 'AI agents request failed');
}

export async function fetchAiRuns(limit = 25): Promise<AiRunListResponse> {
	const params = new URLSearchParams({ limit: String(Math.trunc(limit)) });
	return ApiClient.instance.get<AiRunListResponse>(
		`/api/v1/ai/runs?${params.toString()}`,
		'AI run history request failed'
	);
}

export async function fetchAiSettingsOverview(): Promise<AiSettingsOverviewResponse> {
	return ApiClient.instance.get<AiSettingsOverviewResponse>(
		'/api/v1/ai/settings/overview',
		'AI settings overview request failed'
	);
}

export async function fetchAiProviders(): Promise<AiProviderListResponse> {
	return ApiClient.instance.get<AiProviderListResponse>(
		'/api/v1/ai/providers',
		'AI providers request failed'
	);
}

export async function createAiProvider(request: AiProviderCreateRequest): Promise<AiProviderAccount> {
	return ApiClient.instance.post<AiProviderAccount>(
		'/api/v1/ai/providers',
		request,
		'AI provider create request failed'
	);
}

export async function patchAiProvider(
	providerId: string,
	request: AiProviderPatchRequest
): Promise<AiProviderAccount> {
	return ApiClient.instance.patch<AiProviderAccount>(
		`/api/v1/ai/providers/${encodeURIComponent(providerId)}`,
		request,
		'AI provider update request failed'
	);
}

export async function testAiProvider(providerId: string): Promise<AiProviderCommandResponse> {
	return ApiClient.instance.post<AiProviderCommandResponse>(
		`/api/v1/ai/providers/${encodeURIComponent(providerId)}/test`,
		{},
		'AI provider test request failed'
	);
}

export async function syncAiProviderModels(providerId: string): Promise<AiProviderCommandResponse> {
	return ApiClient.instance.post<AiProviderCommandResponse>(
		`/api/v1/ai/providers/${encodeURIComponent(providerId)}/sync-models`,
		{},
		'AI provider model sync request failed'
	);
}

export async function saveAiProviderConsent(
	providerId: string,
	request: AiProviderConsentRequest
): Promise<AiProviderAccount> {
	return ApiClient.instance.post<AiProviderAccount>(
		`/api/v1/ai/providers/${encodeURIComponent(providerId)}/consent`,
		request,
		'AI provider consent request failed'
	);
}

export async function fetchAiModels(): Promise<AiModelListResponse> {
	return ApiClient.instance.get<AiModelListResponse>('/api/v1/ai/models', 'AI models request failed');
}

export async function putAiModelRoute(
	slot: string,
	request: AiModelRouteUpdateRequest
): Promise<AiModelRoute> {
	return ApiClient.instance.put<AiModelRoute>(
		`/api/v1/ai/model-routes/${encodeURIComponent(slot)}`,
		request,
		'AI model route update request failed'
	);
}

export async function fetchAiPrompts(): Promise<AiPromptListResponse> {
	return ApiClient.instance.get<AiPromptListResponse>('/api/v1/ai/prompts', 'AI prompts request failed');
}

export async function createAiPrompt(request: AiPromptCreateRequest): Promise<AiPromptTemplate> {
	return ApiClient.instance.post<AiPromptTemplate>(
		'/api/v1/ai/prompts',
		request,
		'AI prompt create request failed'
	);
}

export async function createAiPromptVersion(
	promptId: string,
	request: AiPromptVersionCreateRequest
): Promise<AiPromptVersion> {
	return ApiClient.instance.post<AiPromptVersion>(
		`/api/v1/ai/prompts/${encodeURIComponent(promptId)}/versions`,
		request,
		'AI prompt version create request failed'
	);
}

export async function activateAiPromptVersion(
	promptId: string,
	request: AiPromptActivateRequest
): Promise<AiPromptTemplate> {
	return ApiClient.instance.post<AiPromptTemplate>(
		`/api/v1/ai/prompts/${encodeURIComponent(promptId)}/activate`,
		request,
		'AI prompt activation request failed'
	);
}

export async function testAiPrompt(promptId: string, request: AiPromptTestRequest): Promise<AiPromptEvalRun> {
	return ApiClient.instance.post<AiPromptEvalRun>(
		`/api/v1/ai/prompts/${encodeURIComponent(promptId)}/test`,
		request,
		'AI prompt test request failed'
	);
}

export async function requestAiAnswer(request: AiAnswerRequest): Promise<AiAnswerResponse> {
	return ApiClient.instance.post<AiAnswerResponse>('/api/v1/ai/answers', request, 'AI answer request failed');
}

export async function requestAiMeetingPrep(request: AiMeetingPrepRequest): Promise<AiMeetingPrepResponse> {
	return ApiClient.instance.post<AiMeetingPrepResponse>(
		'/api/v1/ai/meeting-prep',
		request,
		'AI meeting prep request failed'
	);
}

File: backend/src/app/router/routes/ai.rs
use super::support::*;

pub(super) fn add_routes(router: Router<AppState>) -> Router<AppState> {
    router
        .route("/api/v1/ai/status", get(get_ai_status))
        .route(
            "/api/v1/ai/settings/overview",
            get(get_ai_settings_overview),
        )
        .route(
            "/api/v1/ai/providers",
            get(get_ai_providers).post(post_ai_provider),
        )
        .route(
            "/api/v1/ai/providers/{provider_id}",
            patch(patch_ai_provider),
        )
        .route(
            "/api/v1/ai/providers/{provider_id}/test",
            post(post_ai_provider_test),
        )
        .route(
            "/api/v1/ai/providers/{provider_id}/sync-models",
            post(post_ai_provider_sync_models),
        )
        .route(
            "/api/v1/ai/providers/{provider_id}/consent",
            post(post_ai_provider_consent),
        )
        .route(
            "/api/v1/ai/provider-auth/start",
            post(post_ai_provider_auth_start),
        )
        .route(
            "/api/v1/ai/provider-auth/{setup_id}",
            get(get_ai_provider_auth_status),
        )
        .route("/api/v1/ai/models", get(get_ai_models))
        .route(
            "/api/v1/ai/models/availability",
            patch(patch_ai_model_availability),
        )
        .route("/api/v1/ai/model-routes/{slot}", put(put_ai_model_route))
        .route(
            "/api/v1/ai/prompts",
            get(get_ai_prompts).post(post_ai_prompt),
        )
        .route(
            "/api/v1/ai/prompts/{prompt_id}/versions",
            post(post_ai_prompt_version),
        )
        .route(
            "/api/v1/ai/prompts/{prompt_id}/activate",
            post(post_ai_prompt_activate),
        )
        .route(
            "/api/v1/ai/prompts/{prompt_id}/test",
            post(post_ai_prompt_test),
        )
        .route("/api/v1/ai/agents", get(get_ai_agents))
        .route("/api/v1/ai/runs", get(get_ai_runs))
        .route("/api/v1/ai/runs/{run_id}", get(get_ai_run))
        .route("/api/v1/ai/answers", post(post_ai_answer))
        .route(
            "/api/v1/ai/task-candidates/refresh",
            post(post_ai_task_candidates_refresh),
        )
        .route("/api/v1/ai/meeting-prep", post(post_ai_meeting_prep))
}
